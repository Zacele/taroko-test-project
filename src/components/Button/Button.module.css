.button {
  padding: 10px 20px;
  font-size: 16px;
  font-weight: 500;
  text-transform: uppercase;
  color: #fff;
  background-color: #1976d2;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  outline: none;
  transition:
    background-color 250ms ease,
    box-shadow 250ms ease;
  display: inline-block;
  text-align: center;
  line-height: 1.5;
  vertical-align: middle;
}

.button:hover {
  background-color: #115293;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.2);
}

.button:active {
  background-color: #0d47a1;
  box-shadow: inset 0 2px 4px 0 rgba(0, 0, 0, 0.2);
}

.button:focus {
  box-shadow: 0 0 0 3px rgba(25, 118, 210, 0.5);
}

.button:disabled {
  background-color: #aaa;
  color: #fff;
  cursor: not-allowed;
  box-shadow: none;
}

.fluid {
  width: 100%;
}

.default {
  background-color: #1976d2;
  color: white;
}

.secondary {
  background-color: #aaa;
  color: white;
}

.success {
  background-color: #4caf50;
  color: white;
}

.danger {
  background-color: #f44336;
  color: white;
}
