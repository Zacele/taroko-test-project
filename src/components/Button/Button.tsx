'use client';
import React from 'react';
import styles from './Button.module.css';
import clsx from 'clsx';

type ButtonVariant = 'default' | 'secondary' | 'success' | 'danger';

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  children: React.ReactNode;
  disabled?: boolean;
  fluid?: boolean;
  variant?: ButtonVariant;
};

const Button: React.FC<ButtonProps> = ({
  children,
  onClick,
  disabled = false,
  fluid = false,
  variant = 'default',
}) => {
  const buttonClass = clsx(styles.button, {
    [styles.fluid]: fluid,
    [styles.default]: variant === 'default',
    [styles.secondary]: variant === 'secondary',
    [styles.success]: variant === 'success',
    [styles.danger]: variant === 'danger',
  });
  return (
    <button className={buttonClass} onClick={onClick} disabled={disabled}>
      {children}
    </button>
  );
};

export default Button;
