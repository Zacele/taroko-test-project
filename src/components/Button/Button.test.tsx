import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Button from './Button';

describe('Button', () => {
  test('renders with children', () => {
    const { asFragment } = render(<Button>Click Me</Button>);
    expect(
      screen.getByRole('button', { name: /click me/i }),
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('handles onClick event', () => {
    const handleClick = jest.fn();
    const { asFragment } = render(
      <Button onClick={handleClick}>Click Me</Button>,
    );
    fireEvent.click(screen.getByText(/click me/i));
    expect(handleClick).toHaveBeenCalledTimes(1);
    expect(asFragment()).toMatchSnapshot();
  });

  test('is disabled when disabled prop is true', () => {
    const { asFragment } = render(<Button disabled>Disabled Button</Button>);
    expect(
      screen.getByRole('button', { name: /disabled button/i }),
    ).toBeDisabled();
    expect(asFragment()).toMatchSnapshot();
  });

  test('has fluid class when fluid prop is true', () => {
    const { asFragment, container } = render(
      <Button fluid>Fluid Button</Button>,
    );
    expect(container.firstChild).toHaveClass('fluid');
    expect(asFragment()).toMatchSnapshot();
  });

  test('applies variant classes correctly', () => {
    const { rerender, asFragment, container } = render(
      <Button variant="default">Default Button</Button>,
    );
    expect(container.firstChild).toHaveClass('default');
    expect(asFragment()).toMatchSnapshot();

    rerender(<Button variant="secondary">Secondary Button</Button>);
    expect(container.firstChild).toHaveClass('secondary');
    expect(asFragment()).toMatchSnapshot();

    rerender(<Button variant="success">Success Button</Button>);
    expect(container.firstChild).toHaveClass('success');
    expect(asFragment()).toMatchSnapshot();

    rerender(<Button variant="danger">Danger Button</Button>);
    expect(container.firstChild).toHaveClass('danger');
    expect(asFragment()).toMatchSnapshot();
  });
});
