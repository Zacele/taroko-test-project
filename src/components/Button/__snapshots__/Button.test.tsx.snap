// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button applies variant classes correctly 1`] = `
<DocumentFragment>
  <button
    class="button default"
  >
    Default Button
  </button>
</DocumentFragment>
`;

exports[`Button applies variant classes correctly 2`] = `
<DocumentFragment>
  <button
    class="button secondary"
  >
    Secondary Button
  </button>
</DocumentFragment>
`;

exports[`Button applies variant classes correctly 3`] = `
<DocumentFragment>
  <button
    class="button success"
  >
    Success Button
  </button>
</DocumentFragment>
`;

exports[`Button applies variant classes correctly 4`] = `
<DocumentFragment>
  <button
    class="button danger"
  >
    Danger Button
  </button>
</DocumentFragment>
`;

exports[`Button handles onClick event 1`] = `
<DocumentFragment>
  <button
    class="button default"
  >
    Click Me
  </button>
</DocumentFragment>
`;

exports[`Button has fluid class when fluid prop is true 1`] = `
<DocumentFragment>
  <button
    class="button fluid default"
  >
    Fluid Button
  </button>
</DocumentFragment>
`;

exports[`Button is disabled when disabled prop is true 1`] = `
<DocumentFragment>
  <button
    class="button default"
    disabled=""
  >
    Disabled Button
  </button>
</DocumentFragment>
`;

exports[`Button renders with children 1`] = `
<DocumentFragment>
  <button
    class="button default"
  >
    Click Me
  </button>
</DocumentFragment>
`;
